services/src/main/java/org/keycloak/protocol/saml/SamlService.java

```diff
@@ -363,32 +350,27 @@ protected Response handleArtifact(String artifact, String relayState) {
                 if (clientArtifactBindingURL == null || clientArtifactBindingURL.isEmpty()) {
                     throw new ConfigurationException("There is no configured artifact resolution service for the client " + client.getClientId());
                 }
-                SOAPMessage returnedMessage = Soap.createMessage().addToBody(doc).call(clientArtifactBindingURL);
-                //handle artifact response content with normal flow.
-                Document soapBodyContents =  Soap.extractSoapMessage(returnedMessage);
-                SAMLDocumentHolder samlDoc = SAML2Request.getSAML2ObjectFromDocument(soapBodyContents);
-                if (!(samlDoc.getSamlObject() instanceof ArtifactResponseType)) {
-                    throw new ProcessingException("Message received from ArtifactResolveService isn't an ArtifactResponseMessage");
-                }
-                ArtifactResponseType art = (ArtifactResponseType) samlDoc.getSamlObject();
 
-                if (art.getAny() instanceof ResponseType) {
-                    Document clientMessage = SAML2Request.convert((ResponseType)art.getAny());
-                    String response = encodeSAMLdocument(clientMessage);
-                    return handleSamlResponse(response, relayState);
-                } else if (art.getAny() instanceof RequestAbstractType) {
-                    Document clientMessage = SAML2Request.convert((RequestAbstractType)art.getAny());
-                    String request = encodeSAMLdocument(clientMessage);
-                    return handleSamlRequest(request, relayState);
-                } else {
-                    throw new ProcessingException("Cannot recognise message contained in ArtifactResponse");
-                }
-            } catch (ProcessingException | ParsingException | ConfigurationException | SOAPException e) {
+                URI clientArtifactBindingURI = new URI(clientArtifactBindingURL);
+
+                ExecutorService executor = session.getProvider(ExecutorsProvider.class).getExecutor("saml-artifact-pool");
+
+                ArtifactResolutionRunnable artifactResolutionRunnable = new ArtifactResolutionRunnable(getBindingType(), asyncResponse, doc, clientArtifactBindingURI, relayState);
+                ScheduledTaskRunner task = new ScheduledTaskRunner(session.getKeycloakSessionFactory(), artifactResolutionRunnable);
+                executor.execute(task);
+
+                // Current transaction must be ignored due to asyncResponse.
+                session.getTransactionManager().rollback();
```

Comment:
You sure about this? this seems scary to me :-)

Reply:
The only way to make it works without warning.... :-)